tls/s2n_psk.c

```diff
@@ -278,55 +306,70 @@ S2N_RESULT s2n_finish_psk_extension(struct s2n_connection *conn)
     return S2N_RESULT_OK;
 }
 
-static S2N_RESULT s2n_psk_set_hmac(struct s2n_psk *psk, s2n_psk_hmac psk_hmac_alg)
+int s2n_psk_set_hmac(struct s2n_psk *psk, s2n_psk_hmac hmac)
 {
-    switch(psk_hmac_alg) {
+    notnull_check(psk);
+    switch(hmac) {
         case S2N_PSK_HMAC_SHA224:     psk->hmac_alg = S2N_HMAC_SHA224; break;
         case S2N_PSK_HMAC_SHA256:     psk->hmac_alg = S2N_HMAC_SHA256; break;
         case S2N_PSK_HMAC_SHA384:     psk->hmac_alg = S2N_HMAC_SHA384; break;
         default:
-            BAIL(S2N_ERR_HMAC_INVALID_ALGORITHM);
+            S2N_ERROR(S2N_ERR_HMAC_INVALID_ALGORITHM);
     }
-    return S2N_RESULT_OK;
+    return S2N_SUCCESS;
 }
 
-int s2n_connection_set_external_psks(struct s2n_connection *conn, struct s2n_external_psk *psk_vec, size_t psk_vec_length)
+int s2n_connection_set_external_psks(struct s2n_connection *conn, struct s2n_psk **psk_list, uint16_t psk_list_len)
 {
-    ENSURE_POSIX_REF(conn);
-    ENSURE_POSIX_REF(psk_vec);
+    notnull_check(conn);
     
     /* Remove all previously-set external psks */
     /* The loop iterates from len to 1 instead of from len-1 to 0 to avoid size_t underflowing */
     for (size_t i = conn->psk_params.psk_list.len; i > 0; i--) {
         size_t i_index = i - 1;
+
         struct s2n_psk *psk = NULL;
         GUARD_AS_POSIX(s2n_array_get(&conn->psk_params.psk_list, i_index, (void**) &psk));
-        ENSURE_POSIX_REF(psk);
+        notnull_check(psk);
+
         if (psk->type == S2N_PSK_TYPE_EXTERNAL) {
-            GUARD(s2n_psk_free(psk));
+            GUARD_AS_POSIX(s2n_psk_wipe(psk));
```

Comment:
Why are we wiping this instead of freeing it?

Reply:
No logic has actually changed here-- I just renamed "s2n_psk_free" to "s2n_psk_wipe", which is more accurate. We need to free all the blobs allocated for the psk, but not the memory allocated for the psk data structure itself. The memory allocated for the psk data structure is owned by the s2n_array.